### ISSUE TITLE
Re-use Connections class in SpatialPooler class

### ISSUE THREAD

**Issue by david-ragazzi** [numenta/nupic#1903]
In order to simplify/uniformizy current spatial and temporal implementations and ease future changes, I think it's very insteresting we implement Connections class in SpatialPooler class. This already was done nicely in htm.java by @cogmission and it seems that works fine. We would move such class into a `connections.py` file and make TemporalMemory and SpatialPooler import it.

Connections class in Java:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/Connections.java

SpatialPooler class in Java re-using connections:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/SpatialPooler.java#L73

TemporalMemory class in Java re-using connections:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/TemporalMemory.java#L67

@chetan51 @cogmission

**Comment by cogmission** [comment]
I can't claim credit for the idea. I just took Chetan's idea and ran with
it. All functions in the SP and TM are completely functional (no state is
maintained in the algorithms). Along with that, the structure (what I will
call the Region when the Network API is complete) will be pulled out of the
Connections class also, afterward these will be separate:

the columns, cells and proximal dendrites

On Wed, Mar 4, 2015 at 7:09 AM, David Ragazzi <notifications@github.com>
wrote:

> In order to simplify/uniformizy current spatial and temporal
> implementations and ease future changes, I think it's very insteresting we
> implement Connections class in SpatialPooler class. This already was done
> nicely in htm.java by @cogmission <https://github.com/cogmission> and it
> seems that works fine. We would move such class into a connections.py
> file and make TemporalMemory and SpatialPooler import it.
>
> Connections class in Java:
>
> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/Connections.java
>
> SpatialPooler class in Java re-using connections:
>
> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/SpatialPooler.java#L73
>
> TemporalMemory class in Java re-using connections:
>
> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/TemporalMemory.java#L67
>
> @chetan51 <https://github.com/chetan51> @cogmission
> <https://github.com/cogmission>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1903>.
>



-- 
*We find it hard to hear what another is saying because of how loudly "who
one is", speaks...*

**Comment by cogmission** [comment]
Hi please excuse the above - gmail decided to send my message before I was
finished editing (ARRRGH!!!!)

I can't claim credit for the idea. I just took Chetan's idea and ran with
it. All functions in the SP and TM are completely functional (no state is
maintained in the algorithms). Along with that, the structure (what I will
call the Region when the Network API is complete) will be pulled out of the
Connections class also, afterward these will all be separate:

1. Structure: the columns, cells and proximal dendrites (all structures
that are not stateful) (these may change if deemed stateful)
2. State: Distal dendrites, pools, synapses etc.
3. Operations: (i.e. Algorithms) SpatialPooler, TemporalMemory etc.

The idea being that a matrix of columns and cells can be reused by the
algorithms and the state can be altered and kept separate.



On Wed, Mar 4, 2015 at 10:51 AM, cogmission <cognitionmission@gmail.com>
wrote:

> I can't claim credit for the idea. I just took Chetan's idea and ran with
> it. All functions in the SP and TM are completely functional (no state is
> maintained in the algorithms). Along with that, the structure (what I will
> call the Region when the Network API is complete) will be pulled out of the
> Connections class also, afterward these will be separate:
>
> the columns, cells and proximal dendrites
>
> On Wed, Mar 4, 2015 at 7:09 AM, David Ragazzi <notifications@github.com>
> wrote:
>
>> In order to simplify/uniformizy current spatial and temporal
>> implementations and ease future changes, I think it's very insteresting we
>> implement Connections class in SpatialPooler class. This already was done
>> nicely in htm.java by @cogmission <https://github.com/cogmission> and it
>> seems that works fine. We would move such class into a connections.py
>> file and make TemporalMemory and SpatialPooler import it.
>>
>> Connections class in Java:
>>
>> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/Connections.java
>>
>> SpatialPooler class in Java re-using connections:
>>
>> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/SpatialPooler.java#L73
>>
>> TemporalMemory class in Java re-using connections:
>>
>> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/research/TemporalMemory.java#L67
>>
>> @chetan51 <https://github.com/chetan51> @cogmission
>> <https://github.com/cogmission>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/numenta/nupic/issues/1903>.
>>
>
>
>
> --
> *We find it hard to hear what another is saying because of how loudly "who
> one is", speaks...*
>



-- 
*We find it hard to hear what another is saying because of how loudly "who
one is", speaks...*

**Comment by david-ragazzi** [comment]
> The idea being that a matrix of columns and cells can be reused by the algorithms and the state can be altered and kept separate.

Yeah.. And even a network state recording/serialization can be limited to Connections class.

**Comment by chetan51** [comment]
+1

**Comment by scottpurdy** [comment]
As the SM and Connections classes are currently implemented, won't this cause a huge performance hit? I wonder if we can get the best of both worlds by implementing the connections class using the sparse matrix implementations? Or can we do some simple optimizations to the connections class to get the same performance as SM?

I suppose first step is to benchmark the impacted SP functions so see what the difference is and we can go from there. I'd like to review the performance benchmarks.

**Comment by chetan51** [comment]
Yes, we would want to benchmark performance before merging a PR that changes SP in this way.

**Comment by david-ragazzi** [comment]
> Or can we do some simple optimizations to the connections class to get the same performance as SM?

This is something. Actually all operations in the PR remains the same, i.e., getting synapses permanences from a collumn and putting into a array, making changes in this array, and putting back the permanences to the column. An optimization in `Connections` could be the use of indices like `self.synapsesPermanencesForColumn = dict()` where the key is the column index, to avoid some loops every time that we need get permanences of a column (currently the PR transverse synapses list of a column and put these permanences into an array).

As my priority is serialization, I still didn't have time to work on benchmark, but I will.

**Comment by breznak** [comment]
what is the status of this? I'd like to see the change for "python-only" Nupic.

**Comment by rhyolight** [comment]
@breznak said:
> what is the status of this? I'd like to see the change for "python-only" Nupic.

@scottpurdy said _"I suppose first step is to benchmark the impacted SP functions so see what the difference is and we can go from there. I'd like to review the performance benchmarks."_ So I suppose that is the first work to be done?

I don't think anyone has actually started working on this yet.
